### DISCUSSION TITLE
Migrate contribution instructions to `CONTRIBUTING.md`.

### DISCUSSION THREAD

**Discussion by kamiradi** [moveit/moveit2#1974] [Ideas]
### Introduction
As suggested in #1922, this discussion is broadly aimed at what the new `CONTRIBUTING.md` file should look like. Correct me if I'm wrong, but the consensus is to migrate contribution instructions to the file rather than having them on the website so that changes can be easily made and linked. 

### Questions to be answered
- [ ] Is this a complete migration, which would entail deprecating contribution instructions from the website?
- [ ] Shall we make sections for each tile on the current website, within `CONTRIBUTION.md`? There will be separate sections within the file for 'good-first-issue',  'enhancing-docs', 'adding new features' etc. Or should these have their own `.md` files?

Please add anything else that I may be missing.

**Comment by tylerjw** [comment]
Thank you for starting this discussion. This should be discussed because if we just went and did that migration, some people would be unpleasantly surprised by it. To the best of my understanding, I'll summarize each side's arguments. Here is the link to the website: https://moveit.ros.org/documentation/contributing/code/

## Website

- The website appears more professional than GitHub
- People who visit the website for this might look at other things on the website (blog posts, etc)
- Building the website took time, and migrating away from it for this doc is a non-zero amount of time we could spend on something else

## CONTRIBUTING.md

- `CONTRIBUTING.md` has become a standard: [Mozilla Science Lab](https://mozillascience.github.io/working-open-workshop/contributing/), [Wikipedia](https://en.wikipedia.org/wiki/Contributing_guidelines)
- Following the standard reduces friction for new contributors as they find the doc where they expect and can reach it without leaving GitHub.
- The doc on the website is more difficult to update (have to learn that website exists, how to build it, and make a PR over there), increasing the cost of keeping it up to date.

## My Position

We must find ways to reduce the difficulty of updating docs and contributing to MoveIt. Following standard conventions for open-source projects on GitHub is one way to move in that direction. Therefore, I am for migrating these docs from the website into a `CONTRIBUTING.md` file.

Tagging @henningkayser @v4hn @davetcoleman because I believe they will have opinions on this.

**Reply by v4hn** [reply]
I think we have bigger issues and either is fine. :)

**Reply by henningkayser** [reply]
As I see it, the content from a typical CONTRIBUTING.md and that on the website are different. The CONTRIBUTING.md usually looks more like a primer on how to contribute, but it rarely includes exact guidelines on style, CI, etc. I think GitHub has a good example: https://github.com/github/docs/blob/main/CONTRIBUTING.md. It's short and precise, but it links to a bunch of more detailed pages in their documentation. Our file could be an improved version of this: https://moveit.ros.org/documentation/contributing/pullrequests/, and then link to our style guides and best practices.

**Reply by tylerjw** [reply]
Here is the contribution guide I most recently read and helped me contribute: https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md

**Comment by kamiradi** [comment]
Thanks for your feedback @tylerjw @henningkayser and @v4hn. 

For now, shall I add the `CONTRIBUTING.md`, which reflects the format of the documents referred to in the discussion?

**Reply by tylerjw** [reply]
If you do that make sure you link to the website section that it currently links to at the top of the  CONTRIBUTING.md file.

**Reply by kamiradi** [reply]
Sounds good @tylerjw, on it!

**Comment by SuyashSapkal** [comment]
Hi @tylerjw ,

I saw that this is still open issue. So wanted to know if I can help for the same. 

I see that there is a requirement for detailed documentation, which is also to be kept short and simple for new contributors.

I am assuming this is the official reference document for new contributors.
[https://moveit.ros.org/documentation/contributing/pullrequests/](https://moveit.ros.org/documentation/contributing/pullrequests/)

I'll try to note down some points which includes what changes I am planning to do.
Maybe you can verify and give me some feedback based on that further.

**Reply by kamiradi** [reply]
Hi @SuyashSapkal 

I am almost on the verge finishing this up and submitting it for feedback. You are welcome to comment on it. I'll update you when I push.

**Reply by SuyashSapkal** [reply]
Hi,
Thank you for the update.

**Reply by kamiradi** [reply]
@SuyashSapkal Apologies for the extremely late reply. I will not be taking this up any further. You are welcome to give it a go.

**Comment by solosangeeth** [comment]
Hello, I would like to work on this. I am new to Open source. It would be great to get some help. Thanks.

In the issue [#1922 ](https://github.com/ros-planning/moveit2/issues/1922) the first task was to create a shorter [first issue template ](https://github.com/ros-planning/moveit2/blob/main/.github/ISSUE_TEMPLATE/first_timers_only.md). Do I need to work on that first?.  As I understand from this discussion, the job is to incorporate necessary details from [https://moveit.ros.org/documentation/contributing/pullrequests/](https://moveit.ros.org/documentation/contributing/pullrequests/) into CONTRIBUTING.md.
